pulsar-common/src/main/java/org/apache/pulsar/common/compression/CompressionCodecNone.java

```diff
@@ -28,8 +29,15 @@ public class CompressionCodecNone implements CompressionCodec {
 
     @Override
     public ByteBuf encode(ByteBuf raw) {
-        // Provides an encoder that simply returns the same uncompressed buffer
-        return raw.retain();
+        int readableBytes = raw.readableBytes();
+        if (readableBytes == raw.capacity()) {
+            return raw.retain();
+        }
+        raw.markReaderIndex();
+        ByteBuf target = PulsarByteBufAllocator.DEFAULT.buffer(readableBytes, readableBytes);
+        target.writeBytes(raw);
+        raw.resetReaderIndex();
+        return target;
     }
```

Comment:
Do you mean: The `MemoryLimitController` manages memory according to the real size of message, but the message is allocated a bigger memory by ` BatchMessageContainerImpl`?

Reply:
Right.

Reply:
LGTM